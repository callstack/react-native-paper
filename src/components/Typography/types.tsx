import type { TypescaleKey } from '../../types';

export type VariantProp<T> =
  | (T extends string ? (string extends T ? never : T) : never)
  | keyof typeof TypescaleKey;
