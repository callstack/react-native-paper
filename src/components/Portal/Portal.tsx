import * as React from 'react';

import type { InternalTheme } from '../../types';

import PortalConsumer from './PortalConsumer';
import PortalHost, { PortalContext, PortalMethods } from './PortalHost';
import {
  Consumer as SettingsConsumer,
  Provider as SettingsProvider,
} from '../../core/settings';
import { ThemeProvider, withInternalTheme } from '../../core/theming';

export type Props = {
  /**
   * Content of the `Portal`.
   */
  children: React.ReactNode;
  /**
   * @optional
   */
  theme: InternalTheme;
};

/**
 * Portal allows rendering a component at a different place in the parent tree.
 * You can use it to render content which should appear above other elements, similar to `Modal`.
 * It requires a [`Portal.Host`](PortalHost) component to be rendered somewhere in the parent tree.
 * Note that if you're using the `Provider` component, this already includes a `Portal.Host`.
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Portal, Text } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <Portal>
 *     <Text>This is rendered at a different place</Text>
 *   </Portal>
 * );
 *
 * export default MyComponent;
 * ```
 */
class Portal extends React.Component<Props> {
  // @component ./PortalHost.tsx
  static Host = PortalHost;

  render() {
    const { children, theme } = this.props;

    return (
      <SettingsConsumer>
        {(settings) => (
          <PortalContext.Consumer>
            {(manager) => (
              <PortalConsumer manager={manager as PortalMethods}>
                <SettingsProvider value={settings}>
                  <ThemeProvider theme={theme}>{children}</ThemeProvider>
                </SettingsProvider>
              </PortalConsumer>
            )}
          </PortalContext.Consumer>
        )}
      </SettingsConsumer>
    );
  }
}

export default withInternalTheme(Portal);
