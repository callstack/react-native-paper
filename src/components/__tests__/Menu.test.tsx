import * as React from 'react';
import { Animated, StyleSheet, View } from 'react-native';

import { act, render, screen, waitFor } from '@testing-library/react-native';

import { getTheme } from '../../core/theming';
import { MD3Elevation } from '../../types';
import Button from '../Button/Button';
import Menu, { ELEVATION_LEVELS_MAP } from '../Menu/Menu';
import Portal from '../Portal/Portal';

const styles = StyleSheet.create({
  contentStyle: {
    borderTopLeftRadius: 0,
    borderTopRightRadius: 0,
  },
});

it('renders visible menu', () => {
  const tree = render(
    <Portal.Host>
      <Menu
        visible
        onDismiss={jest.fn()}
        anchor={<Button mode="outlined">Open menu</Button>}
      >
        <Menu.Item onPress={jest.fn()} title="Undo" />
        <Menu.Item onPress={jest.fn()} title="Redo" />
      </Menu>
    </Portal.Host>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders not visible menu', () => {
  const tree = render(
    <Portal.Host>
      <Menu
        visible={false}
        onDismiss={jest.fn()}
        anchor={<Button mode="outlined">Open menu</Button>}
      >
        <Menu.Item onPress={jest.fn()} title="Undo" />
        <Menu.Item onPress={jest.fn()} title="Redo" />
      </Menu>
    </Portal.Host>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders menu with content styles', () => {
  const tree = render(
    <Portal.Host>
      <Menu
        visible
        onDismiss={jest.fn()}
        anchor={<Button mode="outlined">Open menu</Button>}
        contentStyle={styles.contentStyle}
      >
        <Menu.Item onPress={jest.fn()} title="Undo" />
        <Menu.Item onPress={jest.fn()} title="Redo" />
      </Menu>
    </Portal.Host>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

([0, 1, 2, 3, 4, 5] as MD3Elevation[]).forEach((elevation) =>
  it(`renders menu with background color based on elevation value = ${elevation}`, () => {
    const theme = getTheme(false, true);

    const { getByTestId } = render(
      <Portal.Host>
        <Menu
          visible
          onDismiss={jest.fn()}
          anchor={<Button mode="outlined">Open menu</Button>}
          elevation={elevation}
        >
          <Menu.Item onPress={jest.fn()} title="Undo" />
          <Menu.Item onPress={jest.fn()} title="Redo" />
        </Menu>
      </Portal.Host>
    );

    expect(getByTestId('menu-surface')).toHaveStyle({
      backgroundColor: theme.colors.elevation[ELEVATION_LEVELS_MAP[elevation]],
    });
  })
);

it('uses the default anchorPosition of top', async () => {
  function makeMenu(visible: boolean) {
    return (
      <Portal.Host>
        <Menu
          visible={visible}
          onDismiss={jest.fn()}
          anchor={
            <Button mode="outlined" testID="anchor">
              Open menu
            </Button>
          }
          contentStyle={styles.contentStyle}
        >
          <Menu.Item onPress={jest.fn()} title="Undo" />
          <Menu.Item onPress={jest.fn()} title="Redo" />
        </Menu>
      </Portal.Host>
    );
  }

  render(makeMenu(false));

  jest
    .spyOn(View.prototype, 'measureInWindow')
    .mockImplementation((fn) => fn(100, 100, 80, 32));

  // You must update instead of creating directly and using it because
  // componentDidUpdate isn't called by default in jest. Forcing the update
  // than triggers measureInWindow, which is how Menu decides where to show
  // itself.
  screen.update(makeMenu(true));

  await waitFor(() => {
    const menu = screen.getByTestId('menu-view');
    expect(menu).toHaveStyle({
      position: 'absolute',
      left: 100,
      top: 100,
    });
  });
});

it('respects anchorPosition bottom', async () => {
  function makeMenu(visible: boolean) {
    return (
      <Portal.Host>
        <Menu
          visible={visible}
          onDismiss={jest.fn()}
          anchor={
            <Button mode="outlined" testID="anchor">
              Open menu
            </Button>
          }
          anchorPosition="bottom"
          contentStyle={styles.contentStyle}
        >
          <Menu.Item onPress={jest.fn()} title="Undo" />
          <Menu.Item onPress={jest.fn()} title="Redo" />
        </Menu>
      </Portal.Host>
    );
  }

  render(makeMenu(false));

  jest
    .spyOn(View.prototype, 'measureInWindow')
    .mockImplementation((fn) => fn(100, 100, 80, 32));

  screen.update(makeMenu(true));

  await waitFor(() => {
    const menu = screen.getByTestId('menu-view');
    expect(menu).toHaveStyle({
      position: 'absolute',
      left: 100,
      top: 132,
    });
  });
});

it('animated value changes correctly', () => {
  const value = new Animated.Value(1);
  const { getByTestId } = render(
    <Portal.Host>
      <Menu
        visible
        onDismiss={jest.fn()}
        anchor={<Button mode="outlined">Open menu</Button>}
        testID="menu"
        contentStyle={[{ transform: [{ scale: value }] }]}
      >
        <Menu.Item onPress={jest.fn()} title="Test" />
      </Menu>
    </Portal.Host>
  );
  expect(getByTestId('menu-surface-outer-layer')).toHaveStyle({
    transform: [{ scale: 1 }],
  });

  Animated.timing(value, {
    toValue: 1.5,
    useNativeDriver: false,
    duration: 200,
  }).start();

  act(() => {
    jest.advanceTimersByTime(200);
  });
  expect(getByTestId('menu-surface-outer-layer')).toHaveStyle({
    transform: [{ scale: 1.5 }],
  });
});

it('renders menu with mode "elevated"', () => {
  const { getByTestId } = render(
    <Portal.Host>
      <Menu
        visible
        onDismiss={jest.fn()}
        anchor={<Button mode="outlined">Open menu</Button>}
        mode="elevated"
      >
        <Menu.Item onPress={jest.fn()} title="Undo" />
        <Menu.Item onPress={jest.fn()} title="Redo" />
      </Menu>
    </Portal.Host>
  );

  const menuSurface = getByTestId('menu-surface');

  // Get flattened styles
  const styles = StyleSheet.flatten(menuSurface.props.style);

  expect(styles).toHaveProperty('shadowColor');
  expect(styles).toHaveProperty('shadowOpacity');
});

it('renders menu with mode "flat"', () => {
  const { getByTestId } = render(
    <Portal.Host>
      <Menu
        visible
        onDismiss={jest.fn()}
        anchor={<Button mode="outlined">Open menu</Button>}
        mode="flat"
      >
        <Menu.Item onPress={jest.fn()} title="Undo" />
        <Menu.Item onPress={jest.fn()} title="Redo" />
      </Menu>
    </Portal.Host>
  );

  const menuSurface = getByTestId('menu-surface');

  // Get flattened styles
  const styles = StyleSheet.flatten(menuSurface.props.style);

  expect(styles).not.toHaveProperty('shadowColor');
  expect(styles).not.toHaveProperty('shadowOpacity');
});
