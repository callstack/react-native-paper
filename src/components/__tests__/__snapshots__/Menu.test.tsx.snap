// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders menu with content styles 1`] = `
[
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderRadius": 20,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderColor": "rgba(121, 116, 126, 1)",
              "borderRadius": 20,
              "borderStyle": "solid",
              "borderWidth": 1,
              "flex": undefined,
              "minWidth": 64,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="button-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 19,
                },
              ]
            }
            testID="button"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        undefined,
                        false,
                        {
                          "color": "rgba(103, 80, 164, 1)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ],
                  ]
                }
                testID="button-text"
              >
                Open menu
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessibilityLabel="Close menu"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": "100%",
        }
      }
    />
    <View
      accessibilityViewIsModal={true}
      collapsable={false}
      onAccessibilityEscape={[MockFunction]}
      pointerEvents="box-none"
      style={
        [
          {
            "position": "absolute",
          },
          {
            "left": 8,
            "top": 8,
          },
          undefined,
        ]
      }
      testID="menu-view"
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "transform": [
              {
                "translateX": -0,
              },
              {
                "translateY": -0,
              },
            ],
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgb(243, 237, 246)",
              "borderRadius": 4,
              "borderTopLeftRadius": 0,
              "borderTopRightRadius": 0,
              "opacity": 0,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.15,
              "shadowRadius": 6,
              "transform": [
                {
                  "scaleX": 0,
                },
                {
                  "scaleY": 0,
                },
              ],
            }
          }
          testID="menu-surface-outer-layer"
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "backgroundColor": "rgb(243, 237, 246)",
                "borderRadius": 4,
                "borderTopLeftRadius": 0,
                "borderTopRightRadius": 0,
                "flex": undefined,
                "paddingVertical": 8,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 2,
              }
            }
            testID="menu-surface"
          >
            <View
              accessibilityRole="menuitem"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  false,
                  [
                    {
                      "height": 48,
                      "justifyContent": "center",
                      "maxWidth": 280,
                      "minWidth": 112,
                    },
                    {
                      "paddingHorizontal": 12,
                    },
                    undefined,
                    undefined,
                  ],
                ]
              }
              testID="menu-item"
            >
              <View
                style={
                  [
                    {
                      "flexDirection": "row",
                    },
                    undefined,
                  ]
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    [
                      false,
                      {
                        "justifyContent": "center",
                      },
                      {
                        "maxWidth": 268,
                        "minWidth": 100,
                      },
                      {
                        "marginLeft": 4,
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    maxFontSizeMultiplier={1.5}
                    numberOfLines={1}
                    selectable={false}
                    style={
                      [
                        {
                          "textAlign": "left",
                        },
                        {
                          "color": "rgba(28, 27, 31, 1)",
                          "writingDirection": "ltr",
                        },
                        [
                          {
                            "fontFamily": "System",
                            "fontSize": 16,
                            "fontWeight": "400",
                            "letterSpacing": 0.15,
                            "lineHeight": 24,
                          },
                          [
                            false,
                            {
                              "color": "rgba(28, 27, 31, 1)",
                              "fontFamily": "System",
                              "fontSize": 16,
                              "fontWeight": "400",
                              "letterSpacing": 0.15,
                              "lineHeight": 24,
                            },
                            undefined,
                          ],
                        ],
                      ]
                    }
                    testID="menu-item-title"
                  >
                    Undo
                  </Text>
                </View>
              </View>
            </View>
            <View
              accessibilityRole="menuitem"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  false,
                  [
                    {
                      "height": 48,
                      "justifyContent": "center",
                      "maxWidth": 280,
                      "minWidth": 112,
                    },
                    {
                      "paddingHorizontal": 12,
                    },
                    undefined,
                    undefined,
                  ],
                ]
              }
              testID="menu-item"
            >
              <View
                style={
                  [
                    {
                      "flexDirection": "row",
                    },
                    undefined,
                  ]
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    [
                      false,
                      {
                        "justifyContent": "center",
                      },
                      {
                        "maxWidth": 268,
                        "minWidth": 100,
                      },
                      {
                        "marginLeft": 4,
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    maxFontSizeMultiplier={1.5}
                    numberOfLines={1}
                    selectable={false}
                    style={
                      [
                        {
                          "textAlign": "left",
                        },
                        {
                          "color": "rgba(28, 27, 31, 1)",
                          "writingDirection": "ltr",
                        },
                        [
                          {
                            "fontFamily": "System",
                            "fontSize": 16,
                            "fontWeight": "400",
                            "letterSpacing": 0.15,
                            "lineHeight": 24,
                          },
                          [
                            false,
                            {
                              "color": "rgba(28, 27, 31, 1)",
                              "fontFamily": "System",
                              "fontSize": 16,
                              "fontWeight": "400",
                              "letterSpacing": 0.15,
                              "lineHeight": 24,
                            },
                            undefined,
                          ],
                        ],
                      ]
                    }
                    testID="menu-item-title"
                  >
                    Redo
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
]
`;

exports[`renders not visible menu 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "transparent",
          "borderRadius": 20,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "rgba(121, 116, 126, 1)",
            "borderRadius": 20,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flex": undefined,
            "minWidth": 64,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 19,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      false,
                      {
                        "marginHorizontal": 24,
                        "marginVertical": 10,
                      },
                      undefined,
                      false,
                      {
                        "color": "rgba(103, 80, 164, 1)",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
              testID="button-text"
            >
              Open menu
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders visible menu 1`] = `
[
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderRadius": 20,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderColor": "rgba(121, 116, 126, 1)",
              "borderRadius": 20,
              "borderStyle": "solid",
              "borderWidth": 1,
              "flex": undefined,
              "minWidth": 64,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="button-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 19,
                },
              ]
            }
            testID="button"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        undefined,
                        false,
                        {
                          "color": "rgba(103, 80, 164, 1)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ],
                  ]
                }
                testID="button-text"
              >
                Open menu
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessibilityLabel="Close menu"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": "100%",
        }
      }
    />
    <View
      accessibilityViewIsModal={true}
      collapsable={false}
      onAccessibilityEscape={[MockFunction]}
      pointerEvents="box-none"
      style={
        [
          {
            "position": "absolute",
          },
          {
            "left": 8,
            "top": 8,
          },
          undefined,
        ]
      }
      testID="menu-view"
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "transform": [
              {
                "translateX": -0,
              },
              {
                "translateY": -0,
              },
            ],
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgb(243, 237, 246)",
              "borderRadius": 4,
              "opacity": 0,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.15,
              "shadowRadius": 6,
              "transform": [
                {
                  "scaleX": 0,
                },
                {
                  "scaleY": 0,
                },
              ],
            }
          }
          testID="menu-surface-outer-layer"
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "backgroundColor": "rgb(243, 237, 246)",
                "borderRadius": 4,
                "flex": undefined,
                "paddingVertical": 8,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 2,
              }
            }
            testID="menu-surface"
          >
            <View
              accessibilityRole="menuitem"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  false,
                  [
                    {
                      "height": 48,
                      "justifyContent": "center",
                      "maxWidth": 280,
                      "minWidth": 112,
                    },
                    {
                      "paddingHorizontal": 12,
                    },
                    undefined,
                    undefined,
                  ],
                ]
              }
              testID="menu-item"
            >
              <View
                style={
                  [
                    {
                      "flexDirection": "row",
                    },
                    undefined,
                  ]
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    [
                      false,
                      {
                        "justifyContent": "center",
                      },
                      {
                        "maxWidth": 268,
                        "minWidth": 100,
                      },
                      {
                        "marginLeft": 4,
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    maxFontSizeMultiplier={1.5}
                    numberOfLines={1}
                    selectable={false}
                    style={
                      [
                        {
                          "textAlign": "left",
                        },
                        {
                          "color": "rgba(28, 27, 31, 1)",
                          "writingDirection": "ltr",
                        },
                        [
                          {
                            "fontFamily": "System",
                            "fontSize": 16,
                            "fontWeight": "400",
                            "letterSpacing": 0.15,
                            "lineHeight": 24,
                          },
                          [
                            false,
                            {
                              "color": "rgba(28, 27, 31, 1)",
                              "fontFamily": "System",
                              "fontSize": 16,
                              "fontWeight": "400",
                              "letterSpacing": 0.15,
                              "lineHeight": 24,
                            },
                            undefined,
                          ],
                        ],
                      ]
                    }
                    testID="menu-item-title"
                  >
                    Undo
                  </Text>
                </View>
              </View>
            </View>
            <View
              accessibilityRole="menuitem"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  false,
                  [
                    {
                      "height": 48,
                      "justifyContent": "center",
                      "maxWidth": 280,
                      "minWidth": 112,
                    },
                    {
                      "paddingHorizontal": 12,
                    },
                    undefined,
                    undefined,
                  ],
                ]
              }
              testID="menu-item"
            >
              <View
                style={
                  [
                    {
                      "flexDirection": "row",
                    },
                    undefined,
                  ]
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    [
                      false,
                      {
                        "justifyContent": "center",
                      },
                      {
                        "maxWidth": 268,
                        "minWidth": 100,
                      },
                      {
                        "marginLeft": 4,
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    maxFontSizeMultiplier={1.5}
                    numberOfLines={1}
                    selectable={false}
                    style={
                      [
                        {
                          "textAlign": "left",
                        },
                        {
                          "color": "rgba(28, 27, 31, 1)",
                          "writingDirection": "ltr",
                        },
                        [
                          {
                            "fontFamily": "System",
                            "fontSize": 16,
                            "fontWeight": "400",
                            "letterSpacing": 0.15,
                            "lineHeight": 24,
                          },
                          [
                            false,
                            {
                              "color": "rgba(28, 27, 31, 1)",
                              "fontFamily": "System",
                              "fontSize": 16,
                              "fontWeight": "400",
                              "letterSpacing": 0.15,
                              "lineHeight": 24,
                            },
                            undefined,
                          ],
                        ],
                      ]
                    }
                    testID="menu-item-title"
                  >
                    Redo
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
]
`;
